text/0000-alloc-me-maybe.md

```diff
@@ -0,0 +1,547 @@
+* gecko: semi-task-oriented, best-effort, global, no unwinding
+* server: task-oriented, semi-robust, global, unwinding
+* runtime: whole-system, robust, global, no unwinding
+
+
+
+
+## User Profile: Embedded
+
+Embedded devs are primarily well-aligned with Rust's current strategy. First and foremost, embedded devs just try to *not* dynamically allocate. Memory should ideally all be allocated at startup. In cases where this isn't practical, simply aborting the process is often the next-best choice. Robust embedded systems need to be able to recover from a crash anyway, and aborting is completely fool-proof.
+
+However, sometimes the embedded system needs to process some user-defined tasks with unpredictable allocations, and completely crashing on OOM would be inappropriate. In those cases handling allocation failure is the right solution. In the case of a failure, the entire task usually reports a failure and is torn down. To make this robust, all allocations for a task are usually isolated to a single pool that can easily be torn down. This ensures nothing leaks, and helps avoid fragmentation. The important thing to note is that the embedded developers are ready and willing to take control of all allocations to do this properly.
+
+Some embedded systems do use unwinding, but this is very rare, so it cannot be assumed.
+
+It seems they would be happy to have some system to prevent infallible allocations from ever being used.
+
+
+
+
+## User Profile: Gecko
+
+Gecko is also primarily well-aligned with Rust's current strategy. For the most part, they liberally allocate and are happy to crash on OOM. This is especially palatable now that firefox is multiprocess. However as a *quality of implementation* matter, they occasionally make some subroutines fallible. For instance, it would be unfortunate if a single giant image prevented a page from loading. Similarly, running out of memory while processing a style sheet isn't significantly different from failing to download it.
+
+However in contrast to the embedded case, this isn't done in a particularly principled way. Some parts might be fallible, some might be infallible. Nothing is pooled to isolate tasks. It's just a best-effort affair.
+
+Gecko is built without unwinding.
+
+It seems they would be happy to have some system to prevent infallible allocations from ever being used.
+
+Gecko's need for this API as soon as possible will result in it temporarily forking several of the std collections, which is the primary impetus for this RFC.
+
+
+
+
+
+## User Profile: Server
+
+This represents a commodity server which handles tasks using threads or futures.
+
+Similar to the embedded case, handling allocation failure at the granularity of tasks is ideal for quality-of-implementation purposes. However, unlike embedded development, it isn't considered practical (in terms of cost) to properly take control of everything and ensure allocation failure is handled robustly.
```

Comment:
I already mentioned this, but not as a review comment: there is no evidence that this isn't considered practical. Since this is part of the motivation section, I think it's important to get this right.